src/vehicle.cpp

```diff
@@ -628,6 +626,11 @@ void vehicle::init_state( int init_veh_fuel, int init_veh_status )
         }
     }
 
+    // Additional 50% chance for heavy damage to disabled vehicles
+    if( veh_status == 1 && one_in( 2 ) ) {
+        smash( get_map(), 0.5 );
```

Comment:
I am not sure get_map() is correct to use here, I don't think we can assume that this is called on a vehicle generated on a the main map.

Reply:
What are other variants then?

Reply:
You need to either track down everywhere this function is called and pass in the map the vehicle is being generated on, or adjust this to only edit the vehicle, and not go through the map.

Reply:
There's comment in `smash` function:

https://github.com/CleverRaven/Cataclysm-DDA/blob/0bf16a28f83cef53f674a8043feccfc585dee6e8/src/vehicle.cpp#L966-L973

I guess this should handle all situations.

Reply:
No, because you're saying the vehicle is on the main map when it might not be - `m` will always be `get_map()`, even when the vehicle is not on that map. This only works when the `m` passed to the function is actually the map the vehicle is on.

Reply:
Most of this map stuff is kinda alien to me, so I'm afraid unless someone help me with it I'll have to remove this snippet.

Reply:
> unless someone help me with it I'll have to remove this snippet

Just a suggestion, something like:
```C++
    // Additional 50% chance for heavy damage to disabled vehicles
    if( veh_status == 1 && one_in( 2 ) ) {
        if( get_map().inbounds( project_to<coords::ms>( global_omt_location() ) ) ) {
            // Currently in map bounds - use current map
            smash( get_map(), 0.5 );
        } else {
            // Out of bounds - load separate map
            tinymap tm;
            tm.load( project_to<coords::sm>( global_omt_location() ), false );
            smash( tm, 0.5 );
        }
    }
```

Reply:
This is what I was suggesting:
I am not sure if the above will do anything than the tinymap version because AFAICT, the vehicle location is not set when this function is called.
This is still a little fragile, (e.g. `vehicle::vehicle()` saying the vehicle is always on `get_map()`), but as written that doesn't matter because it won't have a status of 1.  Same thing for the bit in dump, and in game.cpp
Ultimately, I'm not entirely sure of a good way to do this.
```diff
diff --git a/src/dump.cpp b/src/dump.cpp
index 64db4a49b7..7fc1aff11c 100644
--- a/src/dump.cpp
+++ b/src/dump.cpp
@@ -303,8 +303,8 @@ bool game::dump_stats( const std::string &what, dump_mode mode,
             "Aerodynamics coeff", "Rolling coeff", "Static Drag", "Offroad %"
         };
         auto dump = [&rows]( const vproto_id & obj ) {
-            vehicle veh_empty( obj, 0, 0 );
-            vehicle veh_fueled( obj, 100, 0 );
+            vehicle veh_empty( get_map(), obj, 0, 0 );
+            vehicle veh_fueled( get_map(), obj, 100, 0 );
 
             std::vector<std::string> r;
             r.push_back( veh_empty.name );
diff --git a/src/game.cpp b/src/game.cpp
index 815a6224b6..dd48efa4c8 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -995,7 +995,7 @@ vehicle *game::place_vehicle_nearby(
 {
     std::vector<std::string> search_types = omt_search_types;
     if( search_types.empty() ) {
-        vehicle veh( id );
+        vehicle veh( m, id );
         if( veh.max_ground_velocity() == 0 && veh.can_float() ) {
             search_types.emplace_back( "river" );
             search_types.emplace_back( "lake" );
diff --git a/src/map_extras.cpp b/src/map_extras.cpp
index 516f38621f..645d39074a 100644
--- a/src/map_extras.cpp
+++ b/src/map_extras.cpp
@@ -443,7 +443,7 @@ static bool mx_helicopter( map &m, const tripoint &abs_sub )
     auto crashed_hull = VehicleGroup_crashed_helicopters->pick();
 
     // Create the vehicle so we can rotate it and calculate its bounding box, but don't place it on the map.
-    auto veh = std::make_unique<vehicle>( crashed_hull, rng( 1, 33 ), 1 );
+    auto veh = std::make_unique<vehicle>( m, crashed_hull, rng( 1, 33 ), 1 );
 
     veh->turn( dir1 );
 
diff --git a/src/mapgen.cpp b/src/mapgen.cpp
index 277579212a..bac0582d2c 100644
--- a/src/mapgen.cpp
+++ b/src/mapgen.cpp
@@ -6417,7 +6417,7 @@ vehicle *map::add_vehicle( const vproto_id &type, const tripoint &p, const units
     }
 
     // debugmsg("n=%d x=%d y=%d MAPSIZE=%d ^2=%d", nonant, x, y, MAPSIZE, MAPSIZE*MAPSIZE);
-    auto veh = std::make_unique<vehicle>( type, veh_fuel, veh_status );
+    auto veh = std::make_unique<vehicle>( *this, type, veh_fuel, veh_status );
     tripoint p_ms = p;
     veh->sm_pos = ms_to_sm_remain( p_ms );
     veh->pos = p_ms.xy();
diff --git a/src/vehicle.cpp b/src/vehicle.cpp
index ae95828c5a..25c1d84db9 100644
--- a/src/vehicle.cpp
+++ b/src/vehicle.cpp
@@ -264,7 +264,7 @@ units::volume vehicle_stack::max_volume() const
 
 // Vehicle class methods.
 
-vehicle::vehicle( const vproto_id &type_id, int init_veh_fuel,
+vehicle::vehicle( map &placed_on, const vproto_id &type_id, int init_veh_fuel,
                   int init_veh_status ): type( type_id )
 {
     turn_dir = 0_degrees;
@@ -280,13 +280,13 @@ vehicle::vehicle( const vproto_id &type_id, int init_veh_fuel,
         // The game language may have changed after the blueprint was created,
         // so translated the prototype name again.
         name = proto.name.translated();
-        init_state( init_veh_fuel, init_veh_status );
+        init_state( placed_on, init_veh_fuel, init_veh_status );
     }
     precalc_mounts( 0, pivot_rotation[0], pivot_anchor[0] );
     refresh();
 }
 
-vehicle::vehicle() : vehicle( vproto_id() )
+vehicle::vehicle() : vehicle( get_map(), vproto_id() )
 {
     sm_pos = tripoint_zero;
 }
@@ -332,7 +332,7 @@ bool vehicle::remote_controlled( const Character &p ) const
     return false;
 }
 
-void vehicle::init_state( int init_veh_fuel, int init_veh_status )
+void vehicle::init_state( map &placed_on, int init_veh_fuel, int init_veh_status )
 {
     // vehicle parts excluding engines are by default turned off
     for( vehicle_part &pt : parts ) {
@@ -628,15 +628,7 @@ void vehicle::init_state( int init_veh_fuel, int init_veh_status )
 
     // Additional 50% chance for heavy damage to disabled vehicles
     if( veh_status == 1 && one_in( 2 ) ) {
-        if( get_map().inbounds( global_square_location() ) ) {
-            // Currently in map bounds - use current map
-            smash( get_map(), 0.5 );
-        } else {
-            // Out of bounds - load separate map
-            tinymap tm;
-            tm.load( project_to<coords::sm>( global_square_location() ), false );
-            smash( tm, 0.5 );
-        }
+            smash( placed_on, 0.5 );
     }
 
     for( size_t i = 0; i < engines.size(); i++ ) {
diff --git a/src/vehicle.h b/src/vehicle.h
index 14eb329d6e..a5f7f9e776 100644
--- a/src/vehicle.h
+++ b/src/vehicle.h
@@ -785,7 +785,7 @@ class vehicle
         template <typename Func, typename Vehicle>
         static int traverse_vehicle_graph( Vehicle *start_veh, int amount, Func action );
     public:
-        explicit vehicle( const vproto_id &type_id, int init_veh_fuel = -1, int init_veh_status = -1 );
+        vehicle( map &placed_on, const vproto_id &type_id, int init_veh_fuel = -1, int init_veh_status = -1 );
         vehicle();
         vehicle( const vehicle & ) = delete;
         ~vehicle();
@@ -820,7 +820,7 @@ class vehicle
         bool remote_controlled( const Character &p ) const;
 
         // init parts state for randomly generated vehicle
-        void init_state( int init_veh_fuel, int init_veh_status );
+        void init_state( map &placed_on, int init_veh_fuel, int init_veh_status );
 
         // damages all parts of a vehicle by a random amount
         void smash( map &m, float hp_percent_loss_min = 0.1f, float hp_percent_loss_max = 1.2f,
```